sdk/servicebus/service-bus/src/serviceBusClient.ts

```diff
@@ -366,3 +344,86 @@ export class ServiceBusClient {
 function isReceiveMode(mode: any): mode is "peekLock" | "receiveAndDelete" {
   return mode && typeof mode === "string" && (mode === "peekLock" || mode === "receiveAndDelete");
 }
+
+/**
+ * Helper to validate and extract the common arguments from both the get*Receiver() overloads that
+ * have this pattern:
+ *
+ * queue, lockmode, options
+ * topic, subscription, lockmode, options
+ *
+ * @internal
+ * @ignore
+ */
+export function extractReceiverArguments<OptionsT>(
+  connectionStringEntityName: string | undefined,
+  queueOrTopicName1: string,
+  receiveModeOrSubscriptionName2: "peekLock" | "receiveAndDelete" | string,
+  receiveModeOrOptions3: "peekLock" | "receiveAndDelete" | OptionsT,
+  definitelyOptions4: OptionsT
+): {
+  entityPath: string;
+  receiveMode: "peekLock" | "receiveAndDelete";
+  options?: OptionsT;
+} {
+  if (isReceiveMode(receiveModeOrOptions3)) {
+    const topic = queueOrTopicName1;
+    const subscription = receiveModeOrSubscriptionName2;
+
+    validateEntityNamesMatch(connectionStringEntityName, topic, "receiver-topic");
+
+    return {
+      entityPath: `${topic}/Subscriptions/${subscription}`,
+      receiveMode: receiveModeOrOptions3,
+      options: definitelyOptions4
+    };
+  } else if (isReceiveMode(receiveModeOrSubscriptionName2)) {
+    validateEntityNamesMatch(connectionStringEntityName, queueOrTopicName1, "receiver-queue");
+
+    return {
+      entityPath: queueOrTopicName1,
+      receiveMode: receiveModeOrSubscriptionName2,
+      options: receiveModeOrOptions3
+    };
+  } else {
+    throw new TypeError("Invalid receiveMode provided");
+  }
+}
+
+/**
+ * @internal
+ * @ignore
+ */
+export function validateEntityNamesMatch(
+  connectionStringEntityName: string | undefined,
+  queueOrTopicName: string,
+  senderOrReceiverType: "receiver-topic" | "receiver-queue" | "sender"
+) {
+  if (!connectionStringEntityName) {
+    return;
+  }
+
+  if (queueOrTopicName !== connectionStringEntityName) {
+    let entityType;
+    let senderOrReceiver;
+
+    switch (senderOrReceiverType) {
+      case "receiver-queue":
+        entityType = "queue";
+        senderOrReceiver = "Receiver";
+        break;
+      case "receiver-topic":
+        entityType = "topic";
+        senderOrReceiver = "Receiver";
+        break;
+      case "sender":
+        entityType = "queue/topic";
+        senderOrReceiver = "Sender";
+        break;
+    }
+
+    throw new Error(
+      `The connection string for this ServiceBusClient had an EntityPath of '${connectionStringEntityName}' which doesn't match the name of the ${entityType} for this ${senderOrReceiver} ('${queueOrTopicName}')`
```

Comment:
Interestingly, I had a discussion on this with @JoshLove-msft and @ShivangiReja where we came to the conclusion that we shouldn't be throwing an error for this case. There were 2 arguments:
- We generally avoid doing validation on input data if the client can make a successful request to the service and the service can throw a meaningful error. In this case, the subsequent operations on the sender/receiver would result in an error around auth. The reason for this policy is that in the future, the service might change to support a case that the client today deems invalid.
- The corner case of someone copying a connection string for the namespace, and manually appending EntityPath using wrong value or a typo will now get an error even though the SAS details in the connection string are perfectly valid

This is worth opening up a discussion with the team.

Reply:
We discussed this during the DeadLetter meeting a few weeks back. It turned out that the error thrown by the service was sufficiently vague to warrant throwing a more clear error in the client. The error mentioned a signature mismatch, but it wasn't immediately clear what would have caused it. We came to the conclusion that it was better to optimize for the case of the user specifying an entity level connection string and then trying to get a different entity, as opposed to the case where a user incorrectly enters an entity path in the connection string, but has a shared key that should work for other entities. The latter scenario seems far less likely as it would mean a user would have to manually modify the connection string. The former scenario could happen just from copying an entity-level connection string, but then calling GetReceiver/GetSender/etc. on a different entity.